lib/Optimizer/Transforms/ResetBeforeReuse.cpp

```diff
@@ -0,0 +1,174 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "cudaq/Todo.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+
+using namespace mlir;
+
+#define DEBUG_TYPE "reset-before-reuse"
+
+namespace {
+
+static SmallVector<Operation *, 8> sortUsers(const Value::user_range &users,
+                                             const DominanceInfo &dom) {
+  SmallVector<Operation *, 8> orderedUsers;
+
+  for (auto *user : users) {
+    if ([&]() {
+          for (auto iter = orderedUsers.begin(), iterEnd = orderedUsers.end();
+               iter != iterEnd; ++iter) {
+            assert(*iter);
+            if (dom.dominates(user, *iter)) {
+              orderedUsers.insert(iter, user);
+              return false;
+            }
+          }
+          return true;
+        }())
+      orderedUsers.push_back(user);
+  }
+  return orderedUsers;
+}
+
+class QubitResetBeforeReusePass
+    : public cudaq::opt::QubitResetBeforeReuseBase<QubitResetBeforeReusePass> {
+public:
+  QubitResetBeforeReusePass() = default;
+
+  void runOnOperation() override {
+    func::FuncOp funcOp = getOperation();
+    if (!funcOp || funcOp.empty())
```

Comment:
```suggestion
    if (funcOp.empty())
```
The operation has to exist to get here, so we don't need to re-check whether it exists.